src/pipelines/epidemiology/sd_humdata.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, List
+from pandas import DataFrame, concat, merge
+from lib.pipeline import DataSource
+from lib.cast import safe_int_cast
+from lib.time import datetime_isoformat
+from lib.utils import grouped_cumsum
+
+
+class SudanHumdataDataSource(DataSource):
+    def parse_dataframes(
+        self, dataframes: List[DataFrame], aux: Dict[str, DataFrame], **parse_opts
+    ) -> DataFrame:
+
+        # Rename the appropriate columns
+        data = (
+            dataframes[0]
+            .rename(
+                columns={
+                    "Report Date": "date",
+                    "State": "match_string",
+                    "Confirmed Cases": "total_confirmed",
+                }
+            )
+            .drop([0])
+        )
+
+        # The dates in the provided CSV are incorrect for one of the reports.
+        # Replace with report date taken from text of report.
+        data.loc[
+            data["Source"]
+            == "https://reliefweb.int/sites/reliefweb.int/files/resources/Situation%20Report%20-%20Sudan%20-%207%20May%202020.pdf",
+            "date",
+        ] = "5/11/2020"
+
+        data = data.drop(axis=1, columns=["As of Date", "Source"])
+
+        # Remove Abyei PCA, a disputed region with no data shown.
+        data = data[data["match_string"] != "Abyei PCA"]
+
+        # Data source uses different spelling from src/data/iso_3166_2_codes.csv
+        data["match_string"].replace({"Gedaref": "Al Qadarif"}, inplace=True)
+
+        data.date = data.date.apply(lambda x: datetime_isoformat(x, "%m/%d/%Y"))
+
+        # Sudan data includes empty cells where there are no confirmed cases.
+        # These get read in as NaN.  Replace them with zeroes so that the
+        # grouped_diff call to get new confirmed cases works for a state's first
+        # day with a case.
+
+        data["total_confirmed"] = (
```

Comment:
We may want to look into fixing this at a more general level.  Afghanistan also does not report "new_" fields on the first day that a province has a case, I think because diffs with NaN result in NaN.

Reply:
This is tricky, because we can't assume that the first date corresponds to zero cases. Several data sources only started reporting after the first few cases were confirmed.

Reply:
I think it's okay if the data source starts reporting after the first few cases are confirmed, since it looks like (at least from the Sudan data) we don't infer "new_" fields on the first day of a data source.

Do you think that if a data source has a blank entry in a column on a region on a date for which other regions have data for that column, we can usually assume the entry is zero?  Especially if it's a total count entry?

Reply:
No, I don't think that's a fair assumption we can make in a general sense. Different regions may have data coming from completely different places, and maybe a certain health department dropped the ball one day.

An exception to that could be if the data source represents zeros as nulls, and we can be pretty certain of some values being zero (for example, because a certain region reports a single total infection after N days of reporting null values)

Reply:
Good point, yeah I guess it's safer to have any arithmetic involving NaN return NaN (and be surfaced as such, or not surfaced at all, rather than being surfaced as zero).

I can tell that Sudan reports zeros as nulls since the CSV for the first date links to a report at https://reliefweb.int/sites/reliefweb.int/files/resources/Situation%20Report%20-%20Sudan%20-%2030%20Apr%202020.pdf
 which lists the number of states that are affected.  From the report after that onward, all states were affected.